################################################################################
# Brain clock: parse brainSeq Phase I data (from                               #
# DOI: 10.1038/s41593-018-0197-y). We will only pick up brains that don't.     #
# appear in brainSeq Phase II.                                                 #
################################################################################

gene_file <- "../../data/brainSeq_pI/rse_gene_BrainSeq_Phase1_hg19_TopHat2_EnsemblV75.rda"
bs_pII_md_f <- "../../results/parsed/brainseq_pII_metadata.csv"
out_dir <- "../../results/parsed/"

# Load the data
################################################################################
load(gene_file)

# Load BrainSeq Phase II metadata file (generated by brainSeq_pII_parsing.R).
# We will use it to exclude brains from phase I that are already present in
# phase II.
bs_pII_md <- read.csv(bs_pII_md_f, row.names = 1)

# Parse metadata
################################################################################
met_dat <- rse_gene@colData

# Keep only controls (non-controls are schizophrenia patients)
met_dat <- met_dat[met_dat$Dx == "Control", ]
dim(met_dat)

# Keep samples with age higher than 20 yo
met_dat <- met_dat[met_dat$Age >= 20, ]

# Keep brains that don't appear in bs_pII_md
met_dat <- met_dat[!met_dat$BrNum %in% bs_pII_md$individualID, ]

# Create unified metadata dataframe, with same column names as the rest
met_dat_unified <- met_dat[, c("RNum",
                               "RIN",
                               "BrNum",
                               "Region",
                               "Sex",
                               "Race",
                               "Age",
                               "Dx")]
met_dat_unified$RIN <- unlist(met_dat_unified$RIN)
met_dat_unified <- as.data.frame(met_dat_unified)
met_dat_unified$platform <- "TruSeq.v1"
met_dat_unified$libraryPrep <- ""
met_dat_unified$libraryPreparationMethod <- ""
met_dat_unified$runType <- ""
met_dat_unified$readLength <- NA
met_dat_unified$organ <- "brain"
met_dat_unified$assay <- "rnaSeq"
met_dat_unified$exclude <- NA
met_dat_unified$excludeReason <- ""
met_dat_unified$Sex <- gsub("F", "female", met_dat_unified$Sex)
met_dat_unified$Sex <- gsub("M", "male", met_dat_unified$Sex)
met_dat_unified$Race <- gsub("AA", "Black", met_dat_unified$Race)
met_dat_unified$Race <- gsub("AS", "Asian", met_dat_unified$Race)
met_dat_unified$Race <- gsub("CAUC", "White", met_dat_unified$Race)
met_dat_unified$Race <- gsub("HISP", "Hispanic", met_dat_unified$Race)
met_dat_unified$apoeGenotype <- ""
met_dat_unified$pmi <- NA
met_dat_unified$Braak <- NA
met_dat_unified$substudy <- "brainSeq_pI"
met_dat_unified$batch_rna <- NA
met_dat_unified$batch_lib <- ""
met_dat_unified$batch_seq <- NA
met_dat_unified$mmse30_first_ad_dx <- NA
met_dat_unified$mmse30_lv <- NA
met_dat_unified$perturbation <- "none"
met_dat_unified$exper_group <- ""

colnames(met_dat_unified) <- gsub("RNum", "specimenID", colnames(met_dat_unified))
colnames(met_dat_unified) <- gsub("BrNum", "individualID", colnames(met_dat_unified))
colnames(met_dat_unified) <- gsub("Region", "tissue", colnames(met_dat_unified))
colnames(met_dat_unified) <- gsub("Sex", "sex", colnames(met_dat_unified))
colnames(met_dat_unified) <- gsub("Race", "race", colnames(met_dat_unified))
colnames(met_dat_unified) <- gsub("Age", "ageDeath", colnames(met_dat_unified))
colnames(met_dat_unified) <- gsub("Dx", "diagn_4BrainClck", colnames(met_dat_unified))
met_dat_unified <- met_dat_unified[, c("specimenID",
                                       "platform",
                                       "RIN",
                                       "libraryPrep",
                                       "libraryPreparationMethod",
                                       "runType",
                                       "readLength",
                                       "individualID",
                                       "organ",
                                       "tissue",
                                       "assay",
                                       "exclude",
                                       "excludeReason",
                                       "sex",
                                       "race",
                                       "ageDeath",
                                       "apoeGenotype",
                                       "pmi",
                                       "Braak",
                                       "diagn_4BrainClck",
                                       "substudy",
                                       "batch_rna",
                                       "batch_lib",
                                       "batch_seq",
                                       "mmse30_first_ad_dx",
                                       "mmse30_lv",
                                       "perturbation",
                                       "exper_group")]
rownames(met_dat_unified) <- 1:nrow(met_dat_unified)

# Parse counts
################################################################################
counts <- rse_gene@assays$data$counts

# Keep only genes that we kept in metadata
counts <- counts[, colnames(counts) %in% met_dat_unified$specimenID]

write.csv(counts, file = sprintf("%sbrainseq_pI_counts.csv",
                                 out_dir))
write.csv(met_dat_unified, file = sprintf("%sbrainseq_pI_metadata.csv",
                                          out_dir))